0000-plugins.md

```diff
@@ -0,0 +1,98 @@
+- Feature Name: plugins
+- Start Date: 2024-12-02
+- RFC PR: [LemmyNet/rfcs#0000](https://github.com/LemmyNet/rfcs/pull/0008)
+- Lemmy Issue: [LemmyNet/lemmy#0000](https://github.com/LemmyNet/lemmy/issues/3562)
+
+# Summary
+
+Plugins will make Lemmy more flexible, and allow implementing features which are too niche for merging into Lemmy core. It will allow Lemmy to focus on core features, while allowing outside developers to contribute new features without using Rust.
+
+# Motivation
+
+TODO
+
+# Guide-level explanation
+
+## Extism
+
+An open source framework to develop plugins for Rust projects. It supports numerous different languages by compiling them to webassembly which gets called from Rust.
+
+https://extism.org/
+
+# Reference-level explanation
+
+## Plugin Hooks
+
+Lemmy will have to add hooks for specific actions, which can then be used by plugins. In general there are two types of hooks: before an action is written to the database, so it can be rejected or altered. And after it is written to the database, mainly for different types of notifications.
+
+For the initial implementation, the following hooks will be available:
+
+### Before writing to Database
+
+These hooks can be used to reject or alter user actions based on different criteria.
+
+- Post
+    - Create local post
+    - Update local post
+    - Receive federated post
+- Comment
+    - Create local comment
+    - Update local comment
+    - Receive federated comment
+- New vote
```

Comment:
I think this should be made a bit more specific maybe? which hooks exactly will be provided to start with, how are they named, how do you install hooks for them, and what exact structure do the arguments have (since the post structs are pretty large and complex - are they converted to JSON? or passed as a native C-like structure?

Reply:
Added